File: Sources/MagicDiffView/DemoFiles/go/user_model.after.go
package main

import (
	"fmt"
	"regexp"
	"sort"
	"strings"
	"time"
)

type User struct {
	ID        int
	Name      string
	Email     string
	Age       *int // 使用指针表示可选字段
	AvatarURL *string
	CreatedAt time.Time
}

var emailPattern = regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`)

func (u *User) Validate() bool {
	if strings.TrimSpace(u.Name) == "" {
		return false
	}

	if strings.TrimSpace(u.Email) == "" {
		return false
	}

	if !emailPattern.MatchString(u.Email) {
		return false
	}

	if u.Age != nil && (*u.Age < 13 || *u.Age > 120) {
		return false
	}

	return true
}

func (u *User) DisplayName() string {
	if u.Name == "" {
		return "Unknown"
	}
	return u.Name
}

func (u *User) Initials() string {
	parts := strings.Fields(u.Name)
	var initials []rune
	for _, part := range parts {
		if len(part) > 0 {
			initials = append(initials, rune(part[0]))
		}
	}
	return strings.ToUpper(string(initials))
}

func (u *User) String() string {
	str := fmt.Sprintf("User(%d): %s <%s>", u.ID, u.Name, u.Email)
	if u.Age != nil {
		str += fmt.Sprintf(" [%d 岁]", *u.Age)
	}
	return str
}

func (u *User) ToMap() map[string]interface{} {
	result := map[string]interface{}{
		"id":         u.ID,
		"name":       u.Name,
		"email":      u.Email,
		"created_at": u.CreatedAt.Format(time.RFC3339),
	}

	if u.Age != nil {
		result["age"] = *u.Age
	}

	if u.AvatarURL != nil {
		result["avatar_url"] = *u.AvatarURL
	}

	return result
}

type UserManager struct {
	users   []User
	nextID  int
}

func NewUserManager() *UserManager {
	return &UserManager{
		users:  make([]User, 0),
		nextID: 1,
	}
}

func (m *UserManager) AddUser(user User) bool {
	for _, existingUser := range m.users {
		if existingUser.Email == user.Email {
			return false
		}
	}
	m.users = append(m.users, user)
	return true
}

func (m *UserManager) FindUser(id int) *User {
	for i := range m.users {
		if m.users[i].ID == id {
			return &m.users[i]
		}
	}
	return nil
}

func (m *UserManager) FindUserByEmail(email string) *User {
	for i := range m.users {
		if m.users[i].Email == email {
			return &m.users[i]
		}
	}
	return nil
}

func (m *UserManager) RemoveUser(id int) bool {
	for i, user := range m.users {
		if user.ID == id {
			m.users = append(m.users[:i], m.users[i+1:]...)
			return true
		}
	}
	return false
}

func (m *UserManager) UpdateUser(id int, name *string, email *string, age *int) bool {
	for i := range m.users {
		if m.users[i].ID == id {
			if name != nil {
				m.users[i].Name = *name
			}
			if email != nil {
				m.users[i].Email = *email
			}
			if age != nil {
				m.users[i].Age = age
			}
			return true
		}
	}
	return false
}

func (m *UserManager) SortedUsers() []User {
	sorted := make([]User, len(m.users))
	copy(sorted, m.users)

	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].Name < sorted[j].Name
	})

	return sorted
}

func (m *UserManager) AdultUsers() []User {
	var adults []User
	for _, user := range m.users {
		if user.Age != nil && *user.Age >= 18 {
			adults = append(adults, user)
		}
	}
	return adults
}

func (m *UserManager) GetUsersByAgeRange(minAge, maxAge int) []User {
	var result []User
	for _, user := range m.users {
		if user.Age != nil && *user.Age >= minAge && *user.Age <= maxAge {
			result = append(result, user)
		}
	}
	return result
}

func (m *UserManager) TotalCount() int {
	return len(m.users)
}

func (m *UserManager) CreateUser(name, email string, age *int) User {
	user := User{
		ID:        m.nextID,
		Name:      name,
		Email:     email,
		Age:       age,
		CreatedAt: time.Now(),
	}

	m.AddUser(user)
	m.nextID++

	return user
}

File: Sources/MagicDiffView/DemoFiles/python/user_model.after.py
from typing import List, Optional
from datetime import datetime
from dataclasses import dataclass, field
from email.utils import parseaddr
import re


@dataclass
class User:
    id: int
    name: str
    email: str
    age: Optional[int] = None
    avatar_url: Optional[str] = None
    created_at: datetime = field(default_factory=datetime.now)

    def validate(self) -> bool:
        """验证用户数据"""
        if not self.name or not self.name.strip():
            return False

        if not self.email or not self.email.strip():
            return False

        # 验证邮箱格式
        email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        if not re.match(email_pattern, self.email):
            return False

        # 验证年龄
        if self.age is not None and (self.age < 13 or self.age > 120):
            return False

        return True

    def __str__(self) -> str:
        info = f"User({self.id}): {self.name} <{self.email}>"
        if self.age is not None:
            info += f" [{self.age}岁]"
        return info

    def __repr__(self) -> str:
        return self.__str__()

    @property
    def display_name(self) -> str:
        """获取显示名称"""
        return self.name if self.name else "Unknown"

    @property
    def initials(self) -> str:
        """获取姓名首字母"""
        return ''.join(
            word[0].upper()
            for word in self.name.split()
            if word
        )

    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            'id': self.id,
            'name': self.name,
            'email': self.email,
            'age': self.age,
            'avatar_url': self.avatar_url,
            'created_at': self.created_at.isoformat()
        }


class UserManager:
    def __init__(self):
        self.users: List[User] = []
        self._next_id: int = 1

    def add_user(self, user: User) -> bool:
        """添加用户，如果邮箱已存在则返回 False"""
        if any(u.email == user.email for u in self.users):
            return False
        self.users.append(user)
        return True

    def find_user(self, user_id: int) -> Optional[User]:
        """根据 ID 查找用户"""
        return next((u for u in self.users if u.id == user_id), None)

    def find_user_by_email(self, email: str) -> Optional[User]:
        """根据邮箱查找用户"""
        return next((u for u in self.users if u.email == email), None)

    def remove_user(self, user_id: int) -> bool:
        """删除用户，成功返回 True"""
        for i, user in enumerate(self.users):
            if user.id == user_id:
                self.users.pop(i)
                return True
        return False

    def update_user(
        self,
        user_id: int,
        name: Optional[str] = None,
        email: Optional[str] = None,
        age: Optional[int] = None
    ) -> bool:
        """更新用户信息"""
        user = self.find_user(user_id)
        if not user:
            return False

        if name is not None:
            user.name = name
        if email is not None:
            user.email = email
        if age is not None:
            user.age = age

        return True

    @property
    def sorted_users(self) -> List[User]:
        """按姓名排序的用户列表"""
        return sorted(self.users, key=lambda u: u.name)

    @property
    def adult_users(self) -> List[User]:
        """获取成年用户列表"""
        return [u for u in self.users if u.age is not None and u.age >= 18]

    def total_count(self) -> int:
        """获取用户总数"""
        return len(self.users)

    def get_users_by_age_range(self, min_age: int, max_age: int) -> List[User]:
        """获取指定年龄范围的用户"""
        return [
            u for u in self.users
            if u.age is not None and min_age <= u.age <= max_age
        ]

    def create_user(
        self,
        name: str,
        email: str,
        age: Optional[int] = None
    ) -> User:
        """创建并添加新用户"""
        user = User(
            id=self._next_id,
            name=name,
            email=email,
            age=age
        )
        self.add_user(user)
        self._next_id += 1
        return user
